File: backend/app/models/report_self_check.py
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field


class CheckStatus(StrEnum):
    PASS = "pass"
    WARNING = "warning"
    ERROR = "error"


class Confidence(StrEnum):
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


class FindingSeverity(StrEnum):
    WARNING = "warning"
    ERROR = "error"


class Finding(BaseModel):
    severity: FindingSeverity
    title: str
    detail: str
    expected: str | None = None
    actual: str | None = None
    pages: list[int]
    related_fields: list[str]


class EvidenceItem(BaseModel):
    source: str
    page: int | None = None
    label: str | None = None
    value: Any = None


class MissingEvidence(BaseModel):
    label: str
    reason: str
    expected_source: str | None = None


class FieldComparison(BaseModel):
    field: str
    source_a_name: str
    source_a_value: str
    source_a_page: int | None = Field(default=None, gt=0)
    source_b_name: str
    source_b_value: str
    source_b_page: int | None = Field(default=None, gt=0)
    matched: bool
    judgement: str = Field(min_length=1)


class CheckResult(BaseModel):
    check_id: str
    check_name: str
    status: CheckStatus
    confidence: Confidence = Confidence.MEDIUM
    summary: str = ""
    details: dict[str, Any] = Field(default_factory=dict)
    findings: list[Finding] = Field(default_factory=list)
    evidence: list[EvidenceItem] = Field(default_factory=list)
    missing_evidence: list[MissingEvidence] = Field(default_factory=list)


class ReportMeta(BaseModel):
    report_number: str = ""
    sample_number: str = ""
    sample_name: str = ""
    client: str = ""


class SummaryCounts(BaseModel):
    total_checks: int = 0
    pass_count: int = 0
    warning_count: int = 0
    error_count: int = 0


class ReportSelfCheckResult(BaseModel):
    task_id: str
    file_name: str
    overall_status: CheckStatus = CheckStatus.PASS
    report_meta: ReportMeta = Field(default_factory=ReportMeta)
    summary: SummaryCounts = Field(default_factory=SummaryCounts)
    check_results: list[CheckResult] = Field(default_factory=list)

    def refresh_summary(self) -> None:
        pass_count = sum(1 for item in self.check_results if item.status == CheckStatus.PASS)
        warning_count = sum(1 for item in self.check_results if item.status == CheckStatus.WARNING)
        error_count = sum(1 for item in self.check_results if item.status == CheckStatus.ERROR)

        self.summary = SummaryCounts(
            total_checks=len(self.check_results),
            pass_count=pass_count,
            warning_count=warning_count,
            error_count=error_count,
        )
        if error_count:
            self.overall_status = CheckStatus.ERROR
        elif warning_count:
            self.overall_status = CheckStatus.WARNING
        else:
            self.overall_status = CheckStatus.PASS

File: frontend/src/types/reportSelfCheck.ts
export type CheckStatus = 'pass' | 'warning' | 'error';
export type Confidence = 'high' | 'medium' | 'low';
export type RecordReportCheckMode = 'quick' | 'full_codex';

export interface RecordReportCheckOptions {
  record_report_mode: RecordReportCheckMode;
  record_report_concurrency: number;
}

export interface Finding {
  severity: 'warning' | 'error';
  title: string;
  detail: string;
  expected?: string | null;
  actual?: string | null;
  pages: number[];
  related_fields: string[];
}

export interface CheckResult {
  check_id: string;
  check_name: string;
  status: CheckStatus;
  confidence: Confidence;
  summary: string;
  details: Record<string, unknown>;
  findings: Finding[];
  evidence: Array<Record<string, unknown>>;
  missing_evidence: Array<Record<string, unknown>>;
}

export interface ReportSelfCheckResult {
  task_id: string;
  file_name: string;
  ptr_file_name?: string;
  record_file_name?: string;
  report_file_name?: string;
  record_report_mode?: RecordReportCheckMode;
  record_report_concurrency?: number;
  homepage_scope?: Record<string, unknown> | null;
  ptr_report_scope_summary?: Record<string, unknown> | null;
  record_report_summary?: Record<string, unknown> | null;
  overall_status: CheckStatus;
  report_meta: {
    report_number: string;
    sample_number: string;
    sample_name: string;
    client: string;
  };
  summary: {
    total_checks: number;
    pass_count: number;
    warning_count: number;
    error_count: number;
  };
  check_results: CheckResult[];
}

export interface PtrReportCheckResult extends ReportSelfCheckResult {}

export interface ReportSelfCheckTask {
  task_id: string;
  file_name: string;
  ptr_file_name?: string;
  record_file_name?: string;
  report_file_name?: string;
  record_report_mode?: RecordReportCheckMode;
  record_report_concurrency?: number;
  homepage_scope?: Record<string, unknown> | null;
  ptr_report_scope_summary?: Record<string, unknown> | null;
  record_report_summary?: Record<string, unknown> | null;
  status: 'running' | 'completed' | 'error';
  current_check_id: string | null;
  current_check_name: string;
  completed_checks: number;
  total_checks: number;
  logs: string[];
  result: ReportSelfCheckResult | null;
  error: string | null;
}
